crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3972,6 +3973,47 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn nonlocal_name_reference_skips_annotation_only_assignment() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "/src/a.py",
+            "
+            def f():
+                x = 1
+                def g():
+                    // it's pretty weird to have an annotated assignment in a function where the
+                    // name is otherwise not defined; maybe should be an error?
+                    x: int
+                    def h():
+                        y = x
+            ",
+        )?;
+
+        assert_scope_ty(&db, "/src/a.py", &["f", "g", "h"], "y", "Literal[1]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn annotation_only_assignment_transparent_to_local_inference() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "/src/a.py",
+            "
+            x = 1
+            x: int
+            y = x
+            ",
+        )?;
```

Comment:
I guess a question I have here is: so what's the point of this annotation, if our more precise inference takes priority over the declared type? Should we emit a diagnostic warning the user that the declaration will have no effect?

Hmm, and what about something like this? Our precise inference will probably cause the type to be inferred as `list[Literal[1, 2, 3]]`, but that's clearly not what the user wants. And if we infer the precise type, we'll emit a spurious error on the code due to list invariance:

```py
x: list[int] = [1, 2, 3]

def takes_list_of_ints(arg: list[int]) -> None:
    pass

takes_list_of_ints(x)
```

Reply:
This is something I can try to clarify more in the design doc, but I'll try to answer here briefly.

The `x: int` annotation in this particular code sample has no effect, because there is no assignment to `x` anywhere following it. But in general, the effect that it would have would be that any assignment to `x` after `x: int` must assign something that is assignable to `int`, or else there will be an invalid-assignment diagnostic. Without the `x: int`, any subsequent assignment to `x` would be allowed and we'd just happily allow the shadowing to a new inferred type (or infer the union if it's conditional.)

Regarding generics, let me get into that more in the design document.